### PR TITLE
FIX-#3110: preserve partitions shape cache when masking with a slice 

### LINKED ISSUE #3110: When mask partitions with slice it loses its shape cache
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any OS, Ray engine
- **Modin version** (`modin.__version__`): a3ddf2f
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
import modin.pandas as pd

df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})

src_partition = df._query_compiler._modin_frame._partitions[0][0]
masked_partition = src_partition.mask(row_indices=slice(None), col_indices=[0])

assert (
    src_partition._length_cache == masked_partition._length_cache
), f"src: {src_partition._length_cache} | masked: {masked_partition._length_cache}"

```
Output:
```
Traceback (most recent call last):
  File "tst3.py", line 10, in <module>
    ), f"src: {src_partition._length_cache} | masked: {masked_partition._length_cache}"
AssertionError: src: 3 | masked: None
```

### Describe the problem
When masking partition with sliced indices the result partition doesn't preserve the shape cache, however, it still can be computed based on the original partition shape. 

Currently, the new shape is cached only when passed indices have `len` property:
https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/engines/ray/pandas_on_ray/frame/partition.py#L150-L153
For slices, if the original partition already has cached shape, we can recompute new shape with the function: [`modin.pandas.indexing.compute_sliced_len`](https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/pandas/indexing.py#L63)

(Issue related with #3100)

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

This PR adds logic of partition's shape cache re-computing for the result of `.mask(slice)` based on the self shape cache. The common `.mask` logic was moved to the base pandas partition class, the differences of the shape re-computing based on the cache futures of different engines remained to the inheritors.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3110 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [inline modin/engines/base/frame/partition.py:136]: `__copy__` method was missing in the base partition class. Suggested that it wasn't on purpose and added this to be able to use `copy` in the base implementation of the `mask` method.

File: modin/engines/python/pandas_on_python/frame/partition.py
Comment (dchigarev) [inline modin/engines/python/pandas_on_python/frame/partition.py:51]: [note](https://github.com/modin-project/modin/pull/3113/files#r645502745)

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [inline modin/engines/base/frame/partition.py:181]: Since the partition class called `PandasFramePartition` we assume that the underlying partition data is pandas DataFrame

File: modin/pandas/test/test_api.py
Comment (dchigarev) [inline modin/pandas/test/test_api.py:48]: `ignore_modin` here lists sub-modules of `modin.pandas` module, which are not presented in `pandas`, `modin.pandas.indexing` is such type of module. 

Previously this module was never used in the flow of importing `modin.pandas` and so wasn't visible in the `modin.pandas` module scope and hence wasn't listed here. Now we import `modin.pandas.indexing` inside `PandasFramePartition` and so it became visible in `modin.pandas`. That's the reason this line is added.

File: modin/test/test_partition_api.py
Comment (vnlitvinov) [inline modin/test/test_partition_api.py:128]: naming seems wrong... it's a `deserialize` at least :)

**Review by vnlitvinov — CHANGES_REQUESTED** [review]

File: modin/engines/ray/pandas_on_ray/frame/partition.py
Comment (dchigarev) [reply modin/engines/ray/pandas_on_ray/frame/partition.py:221]: We may have length and width cache as futures:
https://github.com/modin-project/modin/blob/653d5b235ffe373b1416fc8049c1080db89bc156/modin/engines/ray/pandas_on_ray/frame/partition.py#L102-L103
In this case, we need to handle them as appropriate: pass them to the remote function where the new shape cache will be recomputed lazily

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [reply modin/engines/base/frame/partition.py:181]: > why we need to cast to pandas.DataFrame here anyway?

`.iloc` may produce scalar in case of single-value result, however, partitions does not support anything more than the pandas DataFrame of Series:
https://github.com/modin-project/modin/blob/653d5b235ffe373b1416fc8049c1080db89bc156/modin/engines/ray/pandas_on_ray/frame/partition.py#L177-L179

Comment (dchigarev) [reply modin/engines/base/frame/partition.py:181]: > each Frame should have an underlying partition class

Yes, it has, but it points to the actually partition class, not to its payload type:
https://github.com/modin-project/modin/blob/653d5b235ffe373b1416fc8049c1080db89bc156/modin/engines/ray/pandas_on_ray/frame/partition_manager.py#L139

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [reply modin/engines/base/frame/partition.py]: removed it. 

we don't need it since started using `copy(obj)`

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [reply modin/engines/base/frame/partition.py]: applied

File: modin/test/test_partition_api.py
Comment (dchigarev) [reply modin/test/test_partition_api.py]: right)

fixed it

File: modin/engines/dask/pandas_on_dask/frame/partition.py
Comment (dchigarev) [reply modin/engines/dask/pandas_on_dask/frame/partition.py:185]: [answer](https://github.com/modin-project/modin/pull/3113/files#r645556269)

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [reply modin/engines/base/frame/partition.py]: moved this logic to the [`is_full_axis_mask`](https://github.com/dchigarev/modin/blob/6dc777efb2e2451c7ceeb6c37c306fb2e6f85f6b/modin/engines/base/frame/partition.py#L154) function

File: modin/engines/base/frame/partition.py
Comment (vnlitvinov) [reply modin/engines/base/frame/partition.py:173]: In that case I suggest checking if it's scalar and only then wrap in the `DataFrame` (less pessimization, more performance).
Also ideally I would like to get `pandas.DataFrame` as a class attribute from somewhere (i.e. not hardcode so it keeps working regardless of backend).

File: modin/pandas/test/test_api.py
Comment (vnlitvinov) [reply modin/pandas/test/test_api.py:48]: we may want to consider splitting `modin.pandas` in two - one that mimics Pandas API and another that is used to implement the first. cc @YarShev

**Comment by vnlitvinov** [comment]
@dchigarev I still think that https://github.com/modin-project/modin/pull/3113#discussion_r645675885 needs addressing before merging...

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [inline modin/engines/base/frame/partition.py:143]: the previous docs were possibly wrong, this function also can take slices or scalar axis labels

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [reply modin/engines/base/frame/partition.py]: >`.iloc` may produce scalar in case of single-value result

found out that it does so only if masked indices are scalars itself, so replaced explicit result conversion to pandas DataFrame by explicit indices conversion to list-like values
```python
>>> df.iloc[0, 0]
1
>>> df.iloc[[0], [0]]
   a
0  1
```

File: modin/engines/base/frame/partition.py
Comment (vnlitvinov) [inline modin/engines/base/frame/partition.py:170]: shouldn't we swap this check and the change to the list in the lines above?..

File: modin/engines/base/frame/partition.py
Comment (vnlitvinov) [inline modin/engines/base/frame/partition.py:187]: maybe unwrap to a full `if..else`?

```suggestion
            if not isinstance(indices, slice):
                return len(indices)
            if not isinstance(previous_cache, int):
                return None
            return compute_sliced_len(indices, previous_cache)
```

looks shorter and is (hopefully) more readable

File: modin/test/test_partition_api.py
Comment (vnlitvinov) [inline modin/test/test_partition_api.py:42]: I would rather we explicitly check for supported engine here (this leaves us with `Python` only for now) and raise an exception in `else` branch so newly added engines do not slip tested in a strange way.

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [reply modin/engines/base/frame/partition.py:170]: I think no, otherwise we need to also add a scalar check to the `is_full_axis_mask`, which is redundant since we can just convert scalar values to the list first.

File: modin/engines/base/frame/partition.py
Comment (dchigarev) [reply modin/engines/base/frame/partition.py]: applied

File: modin/test/test_partition_api.py
Comment (dchigarev) [reply modin/test/test_partition_api.py]: makes sense. applied.

**Review by vnlitvinov — APPROVED** [review]
LGTM, but will merge when CI passes.

File: modin/test/test_partition_api.py
Comment (YarShev) [inline modin/test/test_partition_api.py:39]: @dchigarev , did you use lambda to scatter data under this assumption from dask distributed documentation?
```
Note that it is often better to submit jobs to your workers to have them load the data rather than loading data locally and then scattering it out to them.
```

File: modin/pandas/test/test_api.py
Comment (YarShev) [reply modin/pandas/test/test_api.py:48]: @vnlitvinov , I think we can keep as it is for now.

File: modin/engines/dask/pandas_on_dask/frame/partition.py
Comment (YarShev) [inline modin/engines/dask/pandas_on_dask/frame/partition.py:27]: @dchigarev , why did you import `_get_global_client`? I believe we can use `get_client` that is imported above.

**Review by YarShev — COMMENTED** [review]
Sorry for delayed review.

File: modin/test/test_partition_api.py
Comment (vnlitvinov) [reply modin/test/test_partition_api.py:39]: I think that it's somewhat irrelevant as this is for testing only, and that note is mostly about performance.

File: modin/engines/dask/pandas_on_dask/frame/partition.py
Comment (dchigarev) [reply modin/engines/dask/pandas_on_dask/frame/partition.py:27]: Agree with you. Also, there are more places where this function is used without the need. Created an issue about that (#3133)

**Pull request merged**
